lib/Analysis/Memory/DefinedMemory.cpp

```diff
@@ -730,5 +1070,51 @@ void ReachDFFwk::collapse(DFRegion *R) {
     for (auto Inst : DU->getUnknownInsts())
       DefUse->addUnknownInst(Inst);
   }
+  for (auto &Loc : DefUse->getExplicitAccesses()) {
+    LLVM_DEBUG(dbgs() << "[COLLAPSE] Collapse Explicit Accesses.\n");
+    auto *EM{AT.find(Loc)};
+    if (EM) {
+      if (auto *DFL{dyn_cast<DFLoop>(R)})
+        if (auto *GEP{dyn_cast<GetElementPtrInst>(EM->front())})
+          if (all_of(GEP->indices(), [this, DFL](auto &Idx) {
+                auto *SE{getScalarEvolution()};
+                auto *S{SE->getSCEV(Idx)};
+                return SE->isLoopInvariant(S, DFL->getLoop());
+              }))
+            EM = nullptr;
+          else
+            EM = EM->getTopLevelParent();
+    }
+    MemoryLocationRange ExpLoc = aggregate(R, Loc, this, true).first;
+    MemoryLocationRange NewLoc(Loc);
+    NewLoc.Kind = MemoryLocationRange::LocKind::HINT;
+    if ((ExpLoc.Kind == MemoryLocationRange::LocKind::COLLAPSED &&
+         DefUse->hasDef(ExpLoc)) ||
+        (EM &&
+         any_of(
+             *EM, [EM, &DefUse](auto *Ptr) {
+               return DefUse->hasDef(
+                   MemoryLocation{Ptr, EM->getSize(), EM->getAAInfo()});
+             })))
+      DefUse->addDef(NewLoc);
+    if ((ExpLoc.Kind == MemoryLocationRange::LocKind::COLLAPSED &&
+         DefUse->hasMayDef(ExpLoc)) ||
+        (EM &&
+         any_of(
+             *EM, [EM, &DefUse](auto *Ptr) {
+               return DefUse->hasMayDef(
+                   MemoryLocation{Ptr, EM->getSize(), EM->getAAInfo()});
+             })))
+      DefUse->addMayDef(NewLoc);
+    if ((ExpLoc.Kind == MemoryLocationRange::LocKind::COLLAPSED &&
+         DefUse->hasUse(ExpLoc)) ||
+        (EM &&
+         any_of(
+             *EM, [EM, &DefUse](auto *Ptr) {
+               return DefUse->hasUse(
+                   MemoryLocation{Ptr, EM->getSize(), EM->getAAInfo()});
+             })))
+      DefUse->addUse(NewLoc);

```

Comment:
Убрал дублирование `DEFAULT` и `HINT`, а также убрал проверку для `Def`, иначе он был не консервативен.

```suggestion
    if (!DefUse->hasDef(Loc) &&
        (ExpLoc.Kind == MemoryLocationRange::LocKind::COLLAPSED &&
         DefUse->hasDef(ExpLoc)))
      DefUse->addDef(NewLoc);
    if (!DefUse->hasMayDef(Loc) &&
        ((ExpLoc.Kind == MemoryLocationRange::LocKind::COLLAPSED &&
          DefUse->hasMayDef(ExpLoc)) ||
         (EM && any_of(*EM, [EM, &DefUse](auto *Ptr) {
            return DefUse->hasMayDef(
                MemoryLocation{Ptr, EM->getSize(), EM->getAAInfo()});
          }))))
      DefUse->addMayDef(NewLoc);
    if (!DefUse->hasUse(Loc) &&
        ((ExpLoc.Kind == MemoryLocationRange::LocKind::COLLAPSED &&
          DefUse->hasUse(ExpLoc)) ||
         (EM && any_of(*EM, [EM, &DefUse](auto *Ptr) {
            return DefUse->hasUse(
                MemoryLocation{Ptr, EM->getSize(), EM->getAAInfo()});
          }))))
      DefUse->addUse(NewLoc);
```